### PR TITLE
Batch-parallel K-means and K-medians 

### PR DESCRIPTION
**Pull request by mrfh92**
### Due Diligence

- General:
    - [X]  **base branch** must be `main` for new features, latest release branch (e.g. `release/1.3.x`) for bug fixes
    - [X]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [X] unit tests: all split configurations tested (here: only split=0 allowed)
    - [X] unit tests: multiple dtypes tested
    - [X] documentation updated where needed

### Description

This PR implements a variant of K-Means that can take advantage of Heat data distribution structure. The algorithm is described in the paper https://doi.org/10.1016/j.cie.2020.107023. In essence, the idea is to perform K-means on each chunk of data separately in parallel as a first step. In the second step, the obtained cluster centers are "merged": the set of all process-local cluster centers is collected, and then another application of K-means is used to determine the final "global" cluster centers als cluster centers of the set of "local" cluster centers. 
Of course, this idea generalizes to K-Medians in a straightforward way. To improve scalability of this approach even to very high number of processes, the "merging" can also be done in hierarchical manor.

**Caveat:** This does not necessarily yield the same results as the classical K-Means. 

**Issue/s resolved:** #1287 

### Changes proposed:

- New classes `heat.cluster.BatchParallelKMeans` and  `heat.cluster.BatchParallelKMedians`

### Type of change
new feature

### Memory requirements
TBD

### Performance
Example of ~130GB on up to 8 GPU-nodes of the Terrabyte-Cluster. Left plot shows standard K-means (4 clusters) and right plot shows BatchParallelKmeans (4 and 40 clusters) 

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
@mrfh92 this is great. 

Would it be possible or advisable to make batch-parallel clustering available from within a general `ht.cluster.KMeans` call? I'm thinking something like:

```python
kmeans = ht.cluster.KMeans(n_clusters=k, parallel="batch")
```
where `parallel` can be "batch" or "standard".

**Comment by mrfh92** [comment]
@ClaudiaComito I think K-Means/Medians and Batch-parallel K-Means/Medians should be handled as separate functions since from an algorithmic point of view they are quite different. However, it's possible (and I have done so) to add batch-parallel K-Means/Medias as additional Initialization for classical K-Means/Medians

File: heat/cluster/batchparallelclustering.py [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+
+"""
+Actual classes for batch parallel K-means and K-medians
+"""
+
+
+class BatchParallelKMeans(_BatchParallelKCluster):
+    r"""
+    Batch-parallel K-Means clustering algorithm from Ref. [1].
+    This requires data to be given as DNDarray of shape (n_samples, n_features) with split=0 (i.e. split along the sample axis).
```
Comment (ClaudiaComito) [inline heat/cluster/batchparallelclustering.py:326]:
```suggestion
    The input must be a ``DNDarray`` of shape `(n_samples, n_features)`, with split=0 (i.e. split along the sample axis).
```

File: heat/cluster/batchparallelclustering.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+    The idea of the method is to perform the classical K-Means on each batch of data (i.e. on each process-local chunk of data) individually and in parallel.
```
Comment (ClaudiaComito) [inline heat/cluster/batchparallelclustering.py:327]:
```suggestion
    This method performs K-Means clustering on each batch (i.e. on each process-local chunk) of data individually and in parallel.
```

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Looks great to me @mrfh92 , I only have a few edits to the documentation. Otherwise we can merge.

**Comment by mrfh92** [comment]
@ClaudiaComito I have committed all suggested changes.

**Review by ClaudiaComito — APPROVED** [review]
:clap: @mrfh92 !

**Pull request merged**